Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp

```diff
@@ -391,38 +397,108 @@ class Gx2Fitter {
+              << "\n\t"
+              << "result.lastTrackIndex: " << result.lastTrackIndex << "\n\t"
+              << "result.fittedStates->size(): " << result.fittedStates->size())
+
+          // TODO generalize the update of the currentTrackIndex
+          auto& fittedStates = *result.fittedStates;
+
+          // Mask for the track states. We don't need Smoothed and Filtered
+          TrackStatePropMask mask =
+              ~(TrackStatePropMask::Smoothed | TrackStatePropMask::Filtered);
+
+          // Checks if an existing surface is found during the gx2f-iteration.
+          // If not, a new index will be generated afterwards.
+          // During the first iteration, we will always create a new index.
+          size_t currentTrackIndex = Acts::MultiTrajectoryTraits::kInvalid;
+          if (nUpdate == 0) {
+            ACTS_VERBOSE("   processSurface: nUpdate == 0 decision");
+
+            // Add a <mask> TrackState entry multi trajectory. This allocates
+            // storage for all components, which we will set later.
+            currentTrackIndex =
+                fittedStates.addTrackState(mask, result.lastTrackIndex);
+          } else {
+            ACTS_VERBOSE("   processSurface: nUpdate > 0 decision");
+
+            if (result.lastTrackIndex ==
+                Acts::MultiTrajectoryTraits::kInvalid) {
+              currentTrackIndex = 0;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (kInv->0) = "
+                           << currentTrackIndex);
+            } else if (result.lastTrackIndex < fittedStates.size() - 1) {
+              currentTrackIndex = result.lastTrackIndex + 1;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (n+1) = "
+                           << currentTrackIndex);
+            } else {
+              // Add a <mask> TrackState entry multi trajectory. This allocates
+              // storage for all components, which we will set later.
+              currentTrackIndex =
+                  fittedStates.addTrackState(mask, result.lastTrackIndex);
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (ADD NEW)= "
+                           << currentTrackIndex);
+            }
+          }
 
           // now get track state proxy back
-          auto trackStateProxy = fittedStates.getTrackState(newTrackIndex);
-          trackStateProxy.setReferenceSurface(surface->getSharedPtr());
-          // assign the source link to the track state
-          trackStateProxy.setUncalibratedSourceLink(sourcelink_it->second);
-
-          // Fill the track state
-          trackStateProxy.predicted() = std::move(boundParams.parameters());
+          typename traj_t::TrackStateProxy trackStateProxy =
+              fittedStates.getTrackState(currentTrackIndex);
+
+          // Set the trackStateProxy components with the state from the ongoing
+          // propagation
+          {
+            trackStateProxy.setReferenceSurface(surface->getSharedPtr());
```

Comment:
So this means we're resetting the surface every time? I think we don't expect it to change after the first iterations, correct? Creating the shared pointer here is not free, so maybe it's worth checking if it's set and only resetting it if it isn't?